### PR TITLE
Fix usage of imported consts

### LINKED ISSUE #416: Cannot use imported constants in array type definitions
# Description

## Aim

I'd like to define a number of constants which are used throughout the project in a single `constants.nr` file. These values can then be imported and used while knowing that their usage is always consistent.

A simple example of this is as below
```
// constants.nr
const MAGIC_NUMBER: Field = 3;

// main.nr
mod constants;

fn main(x : pub [Field; constants::MAGIC_NUMBER]) {
  for i in 0..constants::MAGIC_NUMBER {
    constrain x[i] == constants::MAGIC_NUMBER;
  }
}
```

## Expected behavior

Using imported constants should work as if they were defined in the same file.

## Bug

Building this example fails with the error.
```
error: Expected a ] but found ::
  ┌─ /noir-global-value-issue/circuits/src/main.nr:4:34
  │
4 │ fn main(x : pub [Field; constants::MAGIC_NUMBER]) {
```

Pulling out just `MAGIC_NUMBER` and remapping it to a different name to avoid the `::` also fails with the error:

```
thread 'main' panicked at 'index out of bounds: the len is 70 but the index is 18446744073709551615', crates/noirc_frontend/src/node_interner.rs:419:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

# To reproduce

[Repo demonstrating issue](https://github.com/TomAFrench/noir-global-value-issue/tree/master/circuits)

Clone and build.

# Environment

- OS: Ubuntu

For `nargo` users

- _noir-lang/noir_ commit cloned: [20048e7](https://github.com/noir-lang/noir/commit/20048e7ce9b651b0b46083ef5e6674331011eb10)
- Proving backend
  - [x] default
    - Clang: (run `clang --version`)
  - [ ] wasm-base

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves #416 

# Description

Allows Noir developers to use global consts in array annotations. 

## Summary of changes

The main change was switching how we read in fixed array sizes in the parser as well as variables. Both are now read in as Path's and then handled in the resolver. If the path is a singular segment we know we have an individual ident and will treat it as such. Otherwise we lookup our const in the path resolver. If the const being used in the array type annotation is not an int the program fails. 

## Dependency additions / changes

(If applicable.)

## Test additions / changes

I added in the test case from issue #416 into the `nargo/tests/test_data/global_consts` program.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

# Additional context

I will open up an issue to enable array de-sugaring to happen inside of name resolution so that we can use variables when declaring arrays using the array sugar. I also would like to enable constant arrays. Right now constants are limited to ints, bools, and Fields. We should allow devs to import arrays of these types.

### CODE REVIEW

File: crates/nargo/tests/test_data/global_consts/proofs/p.proof [hunk 1]
```diff
@@ -1 +1 @@
...
...4e4ebb27a10af0541eaf744186ea38dab9b788cd6efb832c1ed874bfe21e5401f87b1ebec0712ad13e2ba97a4ca00968e2f18b59bf286abc28554a72d1bade9d5e9eebcc99652d806e62fff9f81c66b48bec216534e8f8c60dba6967a2d9d2e9b532bc756f15736cbe0877b6ae3c6a67fb7aca84abc72bbd155643b1d80d6bfa7e70f8c0e148d384a1442ea11632a7f8dcb96a23080935bb21f3c30ab353ad246d8f514cbb9993915d99875095d4506857c91864e7592567024f4733acd4e9d9d40de4552ba5825d21ee2d36e9133a3d54a047992e53a69525b535bcd1c92cff116405c2454ca0e8455f822a98c8253d9eead8fc872c82a7266adcd0045ef793e1c79b74443c42aefcbe2cdf74b6b332530b93c2be4a54e416834872f76c516dcb65d59578cfcae9d0b886eeceff2eebeecf88246822cc041a36ede1e4dc7110995db01a52a7c73e8956d40e01d4a6e2aa1fed878f3ff6391910f9098988add9919ac48b769335b2807930117de09fd05d05069ec3912a522c7e7ac475a97fc93818c0aadac77f299f2444e39fca06c9e1cc6230cafefa4b0c180d414ef2c84ef9ab1bed1bb5c39eaf1d795b713fd390abb4947c52d13ae708de7acf26ff77ba8f0ac2c0b4af3e1a647996e6472f8c0256408e59879eac4e14b23e140304516c35fa09f20751ebf0f3160ee3c8dddd09b98f16cac47ead7522ca537d080104225eda7826297aa1b10086b810eb07755ae7c74693cb8a22e22d453e6db2094fe45c9f4a31186c5790c359d2345b831421a4c8cb8b04ca30630df593d3e34ed5400b59bf2232372e8c7818146a44e7d52252b00c1ffdb9419f
```
Comment (jfecher) [inline crates/nargo/tests/test_data/global_consts/proofs/p.proof:1]: Some proof and Verifier file changes were committed, lets remove them. May also be a good time to add Verifier.tomls to the gitignore

Comment (vezenovm) [reply crates/nargo/tests/test_data/global_consts/proofs/p.proof]: Should I remove the `Verifier.toml`'s all together right now if we are adding them to the gitignore? I can also just revert the changes to Verifier files for now as well

Comment (jfecher) [reply crates/nargo/tests/test_data/global_consts/proofs/p.proof]: I'd avoid removing them for now, it seems we still fail to prove if they're absent despite them being auto-generated now. E.g. proving fails with `cannot find input file located at ".../noir/crates/nargo/tests/test_data/6/Verifier.toml", run nargo build to generate the missing Prover and/or Verifier toml files` for me when the Verifier.toml is removed.

Comment (vezenovm) [reply crates/nargo/tests/test_data/global_consts/proofs/p.proof]: OK, I just reverted some of the Verifier file changes that were committed

**Review by jfecher — APPROVED** [review]
Looks good. Changing normal variables to have full paths needed to happen eventually - it is also part of the first-class-functions changes

**Comment by TomAFrench** [comment]
I haven't got a good example I can share but I think this might only work for usages inside `main.nr`. i.e. If `foo.nr` defines a global constant then it can be used in `main.nr` but not `bar.nr`.

**Comment by jfecher** [comment]
> I haven't got a good example I can share but I think this might only work for usages inside `main.nr`. i.e. If `foo.nr` defines a global constant then it can be used in `main.nr` but not `bar.nr`.

Is there a specific issue or line of code that you're thinking would break this? It should be handled through the normal importing/path-resolution mechanism for names.

**Comment by TomAFrench** [comment]
It may end up being pebkac in that case. I'll see if I can reproduce it in future.

**Pull request merged**